### ISSUE TITLE
adapter: storage command reconciliation 

### ISSUE THREAD

**Issue by heeringa** [MaterializeInc/materialize#12096]
### Feature request

Follow up of #10549 (much of the description is copied from there and just changed to be applicable to the storage node instead of compute)

Start a design / implementation for reconciliation of storage commands. We expect both the storage controller and storage instance to go down and up. When they do, they need to re-establish the correct set-up. The controller is probably the source of truth for what sources are valid and what their schemas are, but the storage instance is the source of truth for the contents of the sources.

Currently the controller will always attempt to rebuild every source from scratch and populate catalog tables on startup, which causes the storage instance to fail because those sources already exist.

**Comment by jkosh44** [comment]
@antiguru has merged a PR so implement compute command reconciliation here: https://github.com/MaterializeInc/materialize/pull/11103

@antiguru also started to implement storage command reconciliation that was modeled off of compute command reconciliation. I have taken @antiguru's code and updated it so that
1. I have updated it so that it reconciles the `Append` command instead of the `Insert` command.
2. I have updated it to resolve all merge conflicts with main.
That can be found here: https://github.com/MaterializeInc/materialize/pull/12097
